### PR TITLE
add manual seeding for RL experiments

### PR DESCRIPTION
**Pull request by natolambert**
@lvwerra or @younesbelkada is there any case where a model will be initialized randomly? If so, we need to make sure the seeds are set too.

Also, I ran make style and it modified those two lines of setup, so we can change them here or elsewhere.

### CODE REVIEW

**Review by younesbelkada — COMMENTED** [review]
Thanks a lot for this PR and taking care of the randomness! 
I am very much ok for these changes, but I have few questions
1- Does this ensure exact reproducibility? I think you still need to set manual seed before creating each model. Maybe we should document more about this in the documentation
2- Maybe we should set the seed only if the seed `is not None` , but no strong opinion here maybe let's see what @lvwerra says here
Thanks again!

**Comment by natolambert** [comment]
I think we try it and see. It covers most use cases, I just don't really know how model loading etc is handled in transformers. My intuition is they're all pretrained and using torch, so the torch setting should work. Not so hard to check 🤗 (I can double check that)

**Comment by lvwerra** [comment]
As @younesbelkada pointed out we initialize the models before setting up the `PPOTrainer`. I think we could copy the [`set_seed`](https://github.com/huggingface/transformers/blob/820c46a707ddd033975bc3b0549eea200e64c7da/src/transformers/trainer_utils.py#L83) function from transformers (we don't need the TF case) which also takes care of the CUDA case. Then we can use it in the `PPOTrainer` and also expose it to the user.

**Comment by natolambert** [comment]
What is the way the model is initialized? I'm confused because I thought we were fine-tuning existing models. What parts get initialized?

@lvwerra @younesbelkada  -- seems like a gap in my new NLP knowledge :)

**Comment by lvwerra** [comment]
@natolambert it's actually the RL part: the value head is randomly initialized :)

**Comment by younesbelkada** [comment]
As a side note, from what I have discovered, you need to declare the seed before initializing any new module, i.e. in this case before initializing `model` & `ref_model`. Check the example snippet  below:

```python
import torch
import torch.nn as nn

torch.manual_seed(0)
linear_1 = nn.Linear(10, 10)
linear_2 = nn.Linear(10, 10)

# check weights are not the same
assert not torch.allclose(linear_1.weight, linear_2.weight)

torch.manual_seed(0)
linear_1 = nn.Linear(10, 10)
torch.manual_seed(0)
linear_2 = nn.Linear(10, 10)

# check weights are the same
assert torch.allclose(linear_1.weight, linear_2.weight)
```
If we make sure users do this, maybe it will be possible to ensure reproducibility - thus worth documenting it!

**Comment by natolambert** [comment]
I don't think we want all the weights to be the same, no? The initial seed should be enough to maintain reproducibility _across runs_ -- does that make sense?

**Comment by natolambert** [comment]
If we are encountering issues later in this, we could also incorportate `[set_full_determinism](https://github.com/huggingface/transformers/blob/820c46a707ddd033975bc3b0549eea200e64c7da/src/transformers/trainer_utils.py#L58)`

File: trl/trainer/ppo_trainer.py [hunk 1]
```diff
@@ -875,6 +879,17 @@ def create_model_card(self, path: str, model_name: Optional[str] = "TRL Model")
         with open(os.path.join(path, "README.md"), "w", encoding="utf-8") as f:
             f.write(model_card_content)
 
+    def set_seed(seed: int):
```
Comment (younesbelkada) [inline trl/trainer/ppo_trainer.py:882]:
```suggestion
    def set_seed(self, seed: int):
```
This should fix the failing CI tests

File: trl/trainer/ppo_trainer.py [hunk 2]
```diff
@@ -875,6 +879,17 @@ def create_model_card(self, path: str, model_name: Optional[str] = "TRL Model")
...
+        """
+        Helper function for reproducible behavior to set the seed in `random`, `numpy`, `torch` and/or `tf` (if installed).
+        Args:
+            seed (`int`): The seed to set.
+        """
+        random.seed(seed)
+        # np.random.seed(seed)      # if np is added to functionality, add this line
+        torch.manual_seed(seed)
+        torch.cuda.manual_seed_all(seed)
```
Comment (younesbelkada) [inline trl/trainer/ppo_trainer.py:891]:
```suggestion
         if seed is not None and isinstance(seed, int):
              random.seed(seed)
              # np.random.seed(seed)      # if np is added to functionality, add this line
              torch.manual_seed(seed)
              torch.cuda.manual_seed_all(seed)
```

File: trl/trainer/ppo_config.py [hunk 3]
```diff
@@ -90,6 +92,7 @@ def __init__(
         tracker_kwargs: Optional[dict] = {},
         accelerator_kwargs: Optional[dict] = {},
         tracker_project_name: Optional[str] = "trl",
+        seed: Optional[int] = 0,
```
Comment (younesbelkada) [inline trl/trainer/ppo_config.py:95]:
```suggestion
        seed: Optional[int] = None,
```

**Review by younesbelkada — COMMENTED** [review]
Thanks for your work on this @natolambert 
I left few minor comments, IMO we should not force users to set a seed by default but maybe @lvwerra has a different opinion on this, or if you think it's ok to force users to set a seed by default then it's fine for me!

**Comment by natolambert** [comment]
I personally just don't really see an issue with the default seed being 0? It's very normal practice in RL imo. The users for non-research generally will just not touch it?

I don't know how a default seed is particularly different than no seeding.

**Review by younesbelkada — APPROVED** [review]
LGTM thanks for this! 🚀

File: trl/trainer/ppo_trainer.py [hunk 4]
```diff
@@ -875,6 +879,17 @@ def create_model_card(self, path: str, model_name: Optional[str] = "TRL Model")
...
             f.write(model_card_content)
 
+    def set_seed(self, seed: Optional[int]):
+        """
+        Helper function for reproducible behavior to set the seed in `random`, `numpy`, `torch` and/or `tf` (if installed).
+        Args:
+            seed (`int`): The seed to set.
+        """
+        random.seed(seed)
+        # np.random.seed(seed)      # if np is added to functionality, add this line
```
Comment (lvwerra) [inline trl/trainer/ppo_trainer.py:889]: lets maybe add it - don't know if torch or some other library is using numpy somewhere.

File: trl/trainer/ppo_trainer.py [hunk 5]
```diff
@@ -875,6 +879,17 @@ def create_model_card(self, path: str, model_name: Optional[str] = "TRL Model")
         with open(os.path.join(path, "README.md"), "w", encoding="utf-8") as f:
             f.write(model_card_content)
 
+    def set_seed(self, seed: Optional[int]):
```
Comment (lvwerra) [inline trl/trainer/ppo_trainer.py:882]: I would move the function to core.py and expose it to the user via `__init__.py`. So if we wanted we can use it in the script too.

**Review by lvwerra — APPROVED** [review]
Looks good to me, just two minor comments. I think without setting the seed before initializing the value head (i.e. loading the models which is outside the trainer) the training will behave different no matter the internal seeding. That's why I would expose the `set_seed` as a standalone function that's used both internally and externally.

For the default value, the `transformers` Trainer uses 42 :)
